### PR TITLE
SQL-1859: Fix ubuntu issue by giving all enums with variants well defined sizes.

### PR DESCRIPTION
**Pull request by pmeredit**
The DM passes uninitialzed memory to functions like SQLGetEnvAttr, so we need to get the size correct in order to not have tainted results.

### CODE REVIEW

File: odbc/src/api/functions.rs [hunk 1]
```diff
@@ -956,14 +956,15 @@ pub unsafe extern "C" fn SQLDriverConnectW(
             let mongo_connection =
                 odbc_unwrap!(sql_driver_connect(conn, &odbc_uri_string), conn_handle);
             *conn.mongo_connection.write().unwrap() = Some(mongo_connection);
-            let buffer_len = usize::try_from(buffer_length).unwrap();
             // We know the mysql ODBC driver returns SUCCESS if the out_connection_string is NULL.
             // We can also just return SUCCESS if the buffer_len is 0. Likely, users are not
             // expecting to get back a warning when they pass an empty buffer to this, especially
             // given that we only currently support DriverConnectOption::NoPrompt.
```
Comment (pmeredit) [inline odbc/src/api/functions.rs:962]: This is just a defensive programming change that is worth keeping.

File: integration_test/tests/power_bi_tests.rs [hunk 2]
```diff
@@ -144,7 +144,7 @@ mod integration {
             in_connection_string_encoded.push(0);
 
             let str_len_ptr = &mut 0;
```
Comment (pmeredit) [inline integration_test/tests/power_bi_tests.rs:146]: Documentation suggests the buffer size should be 1024 minimum, shouldn't have actually be a problem, however.

File: odbc/src/api/definitions.rs [hunk 3]
```diff
@@ -1,20 +1,23 @@
...
 #[derive(Copy, Clone, Debug, PartialEq, Eq, FromPrimitive)]
+#[repr(i32)]
 pub enum SqlBool {
     False = 0,
     True,
 }
 
 // Environment attributes
 #[derive(Copy, Clone, Debug, PartialEq, Eq, FromPrimitive)]
+#[repr(i32)]
```
Comment (nbagnard) [inline odbc/src/api/definitions.rs:12]: Not that it matter size wise, but in the ODBC headers, the values are all unsigned long. Maybe to match, we should map it to u32 (the spec says i32, but the code in this case seems a better of truth to me)?

File: odbc/src/api/definitions.rs [hunk 4]
```diff
@@ -98,11 +110,13 @@ pub enum RetrieveData {
 pub enum RowOperationPtr {}
 
 #[derive(Clone, Copy, Debug)]
+#[repr(i32)]
 pub enum SimulateCursor {
     NonUnique = 0,
 }
 
 #[derive(Clone, Copy, Debug, FromPrimitive)]
+#[repr(i32)]
```
Comment (nbagnard) [inline odbc/src/api/definitions.rs:119]: The specs says SQLULEN, but the values in the ODBC headers (Windows, unixodbc, iodbc) are unsigned int.
```suggestion
#[repr(u32)]
```

**Review by nbagnard — CHANGES_REQUESTED** [review]
Thank you for figuring out the root cause of the issue and also for adding the buffer length safe guard in SQLDriverConnect!
However, I think that we realized that the correct mapping matters and I went over the spec and ODBC headers for all the attributes below and the mapping you picked for most of them don't match. How did you pick the mapping? 
The ones where you used i32 instead of usize would create the same kind of issue when the DM or application would read it.

File: odbc/src/api/definitions.rs [hunk 5]
```diff
@@ -31,25 +35,29 @@ pub enum CpMatch {
...
 pub enum CursorSensitivity {
     Unspecified = 0,
     Insensitive,
     Sensitive,
 }
 
 #[derive(Clone, Copy, Debug)]
+#[repr(i32)]
```
[the remark this answers is not in the document]
Comment (pmeredit) [reply odbc/src/api/definitions.rs:53]: I think we want to be consistent with the headers over the spec. As far as I am concerned, the headers are the actual spec.

File: odbc/src/api/definitions.rs (hunk 3 above)
Comment (pmeredit) [reply odbc/src/api/definitions.rs:12]: Agreed, the headers are safer to follow.

Comment (pmeredit) [inline odbc/src/api/definitions.rs:12]:
```suggestion
#[repr(u32)]
```

File: odbc/src/api/definitions.rs [hunk 6]
```diff
@@ -66,7 +66,7 @@ pub enum Concurrency {
 }
 
 #[derive(Clone, Copy, Debug, FromPrimitive)]
-#[repr(usize)]
+#[repr(i32)]
```
[the remark this answers is not in the document]
Comment (nbagnard) [reply odbc/src/api/definitions.rs:69]: The negative value is a mistake. Is should be one. 
I checked the headers and all values are prefixed with UL.
```
#define SQL_CURSOR_FORWARD_ONLY         0UL
#define SQL_CURSOR_KEYSET_DRIVEN        1UL
#define SQL_CURSOR_DYNAMIC              2UL
#define SQL_CURSOR_STATIC               3UL
```
Can you switch to u32? Once again, the spec and the headers don't match. Let's stick to the headers.

File: odbc/src/api/definitions.rs [hunk 7]
```diff
@@ -58,6 +66,7 @@ pub enum Concurrency {
 }
 
 #[derive(Clone, Copy, Debug, FromPrimitive)]
+#[repr(i32)]
 pub enum CursorType {
     ForwardOnly = 0,
     KeysetDriven = -1,
```
Comment (nbagnard) [inline odbc/src/api/definitions.rs:72]:
```suggestion
    KeysetDriven = 1
```

**Review by nbagnard — CHANGES_REQUESTED** [review]

File: odbc/src/api/definitions.rs [hunk 8]
```diff
@@ -1,20 +1,23 @@
 use num_derive::FromPrimitive;
 
 #[derive(Copy, Clone, Debug, PartialEq, Eq, FromPrimitive)]
+#[repr(i32)]
```
[the remark this answers is not in the document]
Comment (nbagnard) [reply odbc/src/api/definitions.rs:4]: Created [SQL-1860](https://jira.mongodb.org/browse/SQL-1860) to address this.

**Review by nbagnard — APPROVED** [review]
LGTM!

**Pull request merged**